Update list implementation to work with memory pools

ricfehr3 opened the pull request:
For issue: #151 
:stop_sign: Blocked on PR #132 

PR for updating list implementation to work with the new memory pools. This completely removes malloc from the codebase.

This is blocked on the memory pool PR, and this will be in WIP (needs cleanup and I'd like to test more). However, I think it's at a pretty good spot right now, so I'll get the PR going so others can check it out.

It is a pretty confusing thing to digest. I'm working on more diagrams to explain the entire memory system with the pools and how that relates to lists, but below should explain how this list implementation works a bit. 

<img width="1821" height="1021" alt="Listimplementation" src="https://github.com/user-attachments/assets/283568a0-ca28-4802-a3a8-e1365b995f10" />

drewjohnson2 commented:
I appreciate the chart immensely.

ricfehr3 commented on source/list.c:
```diff
@@ -2,100 +2,181 @@
-    if (index < 0 || index >= list->size) 
-        return NULL;
-    return list->_array[index];
+    int len = 0;
```
Cool, I think I'm going to add that.

MeirGavish commented:
Okay, maybe just wrappers like `is_joker_available_to_shop()`, `get_num_jokers_available_to_shop()`, `set_joker_shop_availability()` to abstract the implementation so the list/bitmap implementation is not too coupled with the usage. 
Also since PR #152 got merged the implementation changed to weight the jokers by rarity by making a temporary array of list indices that match the rarity, so that will need to be accommodated in the bitmap tracker implementation. Maybe making a bitmap for each rarity would work? Or the same implementation could work with bitmap indices but would just need to make sure the joker is available to shop (the index is really set) before putting it in the temporary array.

ricfehr3 commented:
> Okay, maybe just wrappers like `is_joker_available_to_shop()`, `get_num_jokers_available_to_shop()`, `set_joker_shop_availability()` to abstract the implementation so the list/bitmap implementation is not too coupled with the usage. Also since PR #152 got merged the implementation changed to weight the jokers by rarity by making a temporary array of list indices that match the rarity, so that will need to be accommodated in the bitmap tracker implementation. Maybe making a bitmap for each rarity would work? Or the same implementation could work with bitmap indices but would just need to make sure the joker is available to shop (the index is really set) before putting it in the temporary array.

I agree, I'll make sure the interface to the jokers is clear with those functions. Good point. 

When I rebased last night, I noticed that random joker function and I think I've got solution. Though, I'm really hoping to get your opinion on it.

- _Note_, I changed `bitmap` to `bitset` because that is what the data structure is actually called, and I was being silly trying to "map" the "bits" to the array.
- In the function [`game_shop_get_random_joker_idx()`](https://github.com/cellos51/balatro-gba/blob/main/source/game.c#L2520), the operation we are replacing is:
  - _go through the list of Jokers that are not in-play from the `jokers_available_to_shop` list._ 
  - _Go through the list one-by-one and get the JokerId from it's entry_ 
  - _The `JokerId` is an offset into the `JokerInfo` array. We use it to get data from the array_
- The goal of that operation for our use can be also be seen as: 
  - _If a Joker is not in-play, get it's ID, an offset into the `JokerInfo` array._
  
Cool, so each flag in our bitset now represents if a Joker is in-play. So, now instead of a single list to parse through, we need to get the offset at every set flag in the bitset.

There's a couple ways to do that, currently I'm doing it a slow way. I won't explain it too deeply here because I'm going to change it, but here is a permalink to it: https://github.com/ricfehr3/balatro-gba/blob/feature/index-list/source/bitset.c#L89 . Basically, you can use the iterator to get the Joker Id's as you can see here: https://github.com/ricfehr3/balatro-gba/blob/feature/index-list/source/game.c#L2500

The much faster way to do this is to change the iterator method. Instead of `bitset_find_idx_of_nth_set` (still helpful for finding a single index, quickly, but doesn't iterate well.) I plan on just using 3 `int`s. 
- One to track the current word in the array.
- One to track the current bit in the word. 
- One to track the current actual offset (which will be the Joker Id for us).

This operation is quick and simple compared to a `List` iterator, and frees up 40 (or how many jokers we have later) list nodes to be something else that would benefit from a list more.

ricfehr3 commented on source/game.c:
```diff
@@ -2734,19 +2724,24 @@ static int game_shop_get_random_joker_idx()
+    while((joker_id = bitset_itr_next(&itr)) != UNDEFINED)
     {
-        intptr_t joker_id = int_list_get(jokers_available_to_shop, i);
+        //intptr_t joker_id = int_list_get(jokers_available_to_shop, i);
```
delete

ricfehr3 commented on source/game.c:
```diff
@@ -254,14 +259,45 @@ bool is_shortcut_joker_active(void)
     return straight_and_flush_size;
 }
 
+static inline void _set_shop_joker_avail(int joker_id, bool avail)
```
@MeirGavish Here are the interfaces to the bitmap we talked about. There is one use of a `BitmapItr` below, but it's in another `Joker` function, so I'm happy with it.

ricfehr3 commented:
Moving this out of draft and opening it up for review.

Please do not merge yet, I'm writing tests but hopefully I won't be changing much with what I have here. 

I was going to make a macro to iterate through the `List` and `Bitset`, but I found that the iterator pattern worked nicely and I didn't feel a need to make it more complicated. It's a pretty simple interface:

```c
    ListItr itr = list_itr_new(&list);
    ListNode* ln;

    while((ln = list_itr_next(&itr)))
    {
        MyObject* obj = ln->data; 
        // do stuff
        ...
    }
```

MeirGavish commented:
> Finishing this up to unblock new work. Mainly #207

@ricfehr3 You meant to unblock #158?

MeirGavish commented on source/game.c:
```diff
@@ -2767,39 +2761,33 @@ static int game_shop_get_random_joker_idx()
 
     for (int i = 0; i < MAX_SHOP_JOKERS; i++)
     {
         intptr_t joker_id = 0;
```
```suggestion
        int joker_id = 0;
```

I guess this should be changed from inptr_t now as there's no reason for it

MeirGavish commented on source/game.c:
```diff
@@ -366,47 +404,7 @@ bool is_joker_owned(int joker_id)
-
-    list_remove_by_idx(jokers, joker_idx);        
+    list_push_back(&_active_jokers_list, joker_object);
 }
```
Here too, keep this wrapper, we will need it...

MeirGavish commented on source/game.c:
```diff
@@ -2893,38 +2891,43 @@ static void jokers_sel_row_on_selection_changed(SelectionGrid *selection_grid,
+        if(joker_object != NULL) {
+            erase_price_under_sprite_object(joker_object->sprite_object);
+            sprite_object_set_focus(joker_object->sprite_object, false);
+        }
```
```suggestion
        if (joker_object != NULL) 
        {
            erase_price_under_sprite_object(joker_object->sprite_object);
            sprite_object_set_focus(joker_object->sprite_object, false);
        }
```

Allman style, we should reformat everything and add clang enforcement.

MeirGavish commented on source/game.c:
```diff
@@ -2893,38 +2891,43 @@ static void jokers_sel_row_on_selection_changed(SelectionGrid *selection_grid,
-        JokerObject* joker_object = list_get(jokers, prev_selection->x);
-        erase_price_under_sprite_object(joker_object->sprite_object);
-        sprite_object_set_focus(joker_object->sprite_object, false);
+        JokerObject* joker_object = (JokerObject*)list_get_at_idx(&_active_jokers_list, prev_selection->x);
```
Originally I wasn't worried about this since I said the list wouldn't get very long but now I do wonder how significant the increase in runtime is here.

ricfehr3 replied:
Here I think we are good. Since the number of "owned" jokers is always going to be within a reasonable value, and this operation only happens in the shop on a sale. 

However, I haven't tested performance. No easy way to, as far as I know?

MeirGavish replied:
Yeah, what worried me is that this is called for every move of the selection when highlighting different jokers in the shop.
I tried testing performance by activating the HBLANK affine background effect on the rest of the game by simply removing the `IRQ_HBLANK` toggle in `affine_background_change_background()`. Doing that creates other performance artifacts but I don't see anything when selecting shop jokers so it seems to be fine.

ricfehr3 commented on source/game.c:
```diff
@@ -228,9 +230,12 @@ static int selection_y = 0;
-static List *jokers = NULL;
-static List *discarded_jokers = NULL;
-static List *jokers_available_to_shop; // List of joker IDs
+static List _active_jokers_list;
```
I agree, that was annoying me too. I'm switching it to `_owned_jokers_list`

ricfehr3 commented on source/game.c:
```diff
@@ -366,47 +404,7 @@ bool is_joker_owned(int joker_id)
-    if (joker_object->joker->id == SHORTCUT_JOKER_ID) 
-    {
-        shortcut_joker_count++;
-    }
```
Readded. Sorry, sloppy rebase work there.

ricfehr3 commented on source/game.c:
```diff
@@ -2893,38 +2891,43 @@ static void jokers_sel_row_on_selection_changed(SelectionGrid *selection_grid,
 }
 
-void game_sell_joker(int joker_idx)
+void game_sell_joker(int joker_id)
```
This does need to be `joker_idx`. It's a bit confusing as the Joker's ID is also it's index into the bitset. The ideas are different and shouldn't be treated the same (as I did here), but it does work, somewhat unfortunately. 

Definitely moving back to `joker_idx`, tho.

MeirGavish replied:
But here it's the index of the joker in the shop list, which is different to its ID obtained through `joker_object->joker->id` and is the actual value passed to the bitset.

ricfehr3 replied:
@MeirGavish Fixed it. I agree with your reasoning 100%. It's called as the idx in the game shop itself. Makes so much more sense now, haha.

ricfehr3 commented on source/game.c:
```diff
@@ -2097,10 +2088,10 @@ static void cards_in_hand_update_loop(bool* discarded_card, int* played_selectio
     {
         (*iteration_start)++;
-        JokerObject *joker = list_get(jokers, k);
+        JokerObject *joker = list_get_at_idx(&_active_jokers_list, k);
```
I agree. It's hard to test... I might need some ideas to make sure I'm not breaking anything, but the iterator replacement seems to be working great.

MeirGavish commented:
I know you might be in the middle of changing things but I noticed some things when running this I thought I should mention.
Owned jokers can reappear in the shop, you can test it by giving yourself a lot of `STARTING_MONEY` and rerolling a lot.

<img width="240" height="160" alt="balatro-gba-5" src="https://github.com/user-attachments/assets/95d1b5a1-3fdb-460c-9ae8-d4690cd03f86" />
<img width="240" height="160" alt="balatro-gba-6" src="https://github.com/user-attachments/assets/2db6fb27-0111-4aff-8635-b988f58d35f2" />


Also the motion effects when rerolling are bugged.

https://github.com/user-attachments/assets/f8283830-ac35-4b2f-a48c-801ac561e2b7

ricfehr3 commented:
@MeirGavish You beat me to it, ha. I'll have to come back to this tomorrow. I thought I had fixed it, but I guess not. I'll need to sleep before looking at this again, haha. Cheers

ricfehr3 commented:
@MeirGavish Alrighty, after a sleeping and working, I found the issue right away (_I think_). Good catch on that. I'm pretty happy with this, besides how to test the jokers exactly.

MeirGavish commented:
Nice! I'll try to take another look tomorrow or during the weekend

MeirGavish commented on source/game.c:
```diff
@@ -2893,38 +2928,45 @@ static void jokers_sel_row_on_selection_changed(SelectionGrid *selection_grid,
 
-    remove_held_joker(joker_idx);
-    int_list_append(jokers_available_to_shop, (intptr_t)joker_object->joker->id);
+    list_remove_at_idx(&_owned_jokers_list, joker_id);
```
This should call `remove_held_joker()` so it calls everything it encapsulates such as the Four Fingers/Shortcut logic

MeirGavish commented on source/game.c:
```diff
@@ -2893,38 +2928,45 @@ static void jokers_sel_row_on_selection_changed(SelectionGrid *selection_grid,
-    int_list_append(jokers_available_to_shop, (intptr_t)joker_object->joker->id);
+    list_remove_at_idx(&_owned_jokers_list, joker_id);
+    
+    _set_shop_joker_avail(joker_object->joker->id, true);
```
This should call remove_held_joker() so it calls everything it encapsulates such as the Four Fingers/Shortcut logic, at the current state selling these jokers will not deactivate them. I'm also thinking the call to `_set_shop_joker_avail` should be moved there.

ricfehr3 replied:
Done, I did modify the `remove_held_joker()` function to take the joker_object and the index in the owned joker list to prevent traversing the list more than once.

MeirGavish commented on include/list.h:
```diff
@@ -1,27 +1,179 @@
+
+/**
+ * Create a new list.
+ *
```
```suggestion
 *
 * The list must be freed with @ref list_destroy
 * While memory is not allocated for the list itself, the list does allocate memory for each element.
 *
```

Add some documentation instructing the list must freed.

MeirGavish commented on include/list.h:
```diff
@@ -1,27 +1,179 @@
+     * @brief The current node in the list
+     */
+    ListNode* current_node;
+} ListItr;
```
The documentation for each field here feels a little excessive, I think all of these are obvious from the name and to whoever knows what a linked list looks like. Maybe it's better to avoid excessive comments and let the code be self-documenting but I will not insist on this if you really prefer it this way.

ricfehr3 replied:
@MeirGavish This was a sneaky attempt from me to introduce Doxygen to this project. The code should be self documenting 100%, and I feel that it does a good job at that right now. But Doxygen on top of it enforces a consistent style and Imo is much easier to parse through once it's in a webpage. 

Totally not necessary though, however it sure is nice to look at once it's done. I'll let you decide on what you wanna do. I personally would like to have the whole repo documented this way (only header files look like this), and I would like to do the work for it. But I'll let you stew on it and decide. I'm good either way.

Here's some pictures of what it looks like with these changes. You can link every variable this way as well.

<img width="1860" height="970" alt="image" src="https://github.com/user-attachments/assets/48c3b311-9d15-4700-879f-41dfccb6934e" />

<img width="838" height="760" alt="image" src="https://github.com/user-attachments/assets/dc97ad54-9d7c-4c7b-9dc8-c867b8cf0c8e" />

MeirGavish commented on include/list.h:
```diff
@@ -1,27 +1,179 @@
+ *
+ * @return A pointer to the next @ref ListNode if valid, otherwise return NULL.
+ */
+ListNode* list_itr_next(ListItr* itr);
```
This is more of a design comment that applies to several places in the code but I thought this is the best place to put it.
I don't think the only way to iterate over the list should give the user access to the nodes, in most usages the list user just wants access to the data and gets it from the node so this iterator should return the data. This will remove code duplication for accessing the data, decouple iteration from the list implementation, and will be safer since the list user doesn't get access to an internal structure, and preserve the list's responsibility for the nodes.
The one place where the node is needed is when removing during iteration. For that I actually think it would be best to still encapsulate it within the list iterator API with a function such as `list_iter_remove_prev()` that would prevent the user from somehow removing a node that is further ahead in the iteration, potentially breaking it.
Barring that, at the very least provide separate functions - `list_itr_next()` that returns the data and `list_itr_next_node()` that returns the node and in the documentation discourage usage of the latter over the former.
I think we may end up breaking this separation to some extend and provide access to the nodes for performance but any function that provides access to nodes should have a preferred alternative.

ricfehr3 replied:
Holy smokes, that actually was a really clean refactor. Good call. The access to the `List` internals are well abstracted now and the cumbersome handling of the list nodes is only in `list.c` now.

MeirGavish commented on source/game.c:
```diff
@@ -346,16 +380,17 @@ int get_scored_card_index(void)
-{
-    return jokers;
+List* get_jokers_list(void) {
+    return &_owned_jokers_list;
```
```suggestion
List* get_jokers_list(void) 
{
    return &_owned_jokers_list;
```

Allman

MeirGavish commented on source/game.c:
```diff
@@ -346,16 +380,17 @@ int get_scored_card_index(void)
-{
-    for (int k = 0; k < list_get_size(jokers); k++)
+bool is_joker_owned(int joker_id) {
+    ListItr itr = list_itr_new(&_owned_jokers_list);
```
```suggestion
bool is_joker_owned(int joker_id) 
{
    ListItr itr = list_itr_new(&_owned_jokers_list);
```

Allman

MeirGavish commented on source/list.c:
```diff
@@ -1,121 +1,170 @@
-    if (list == NULL)
+    if(n >= list_get_len(list) || n < 0) return NULL;
+
+    int len = 0;
```
Maybe a more correct name would be `curr_idx` since this isn't really a length of anything.

MeirGavish reviewed (changes requested):
These are good improvements, well done! I also tested it and it seems the bugs we've seen really are fixed. There are still a few things that need fixing and this time I gave a closer look to the bitset and list APIs.

One more thing, you're the one who first started adding unit tests so huge props for that! I want to keep that habit of adding tests to added code when possible and I've been thinking that now that the bitset is its own separate module (which is also great), it should also have tests along with the pool and list.

I think I'm pretty much done reviewing this PR, but I will want to give game.c another pass since it's a big file with many changes.

ricfehr3 commented on source/game.c:
```diff
@@ -2095,12 +2123,13 @@ static void cards_in_hand_update_loop(bool* discarded_card, int* played_selectio
 
 // returns true if a joker was scored, false otherwise
-static bool check_and_score_joker_for_event(int* iteration_start, Card* played_card, enum JokerEvent joker_event)
+static bool check_and_score_joker_for_event(ListItr* itr, Card* played_card, enum JokerEvent joker_event)
```
Love it, adding.

ricfehr3 commented on source/joker_effects.c:
```diff
@@ -859,8 +863,8 @@ static JokerEffect brainstorm_joker_effect(Joker *joker, Card *scored_card, enum
     }
 
-    List* jokers = get_jokers();
-    JokerObject* first_joker = list_get(jokers, 0);
```
That is fair. It will return NULL and it makes more sense than directly accessing the internals of the list.

ricfehr3 commented on include/bitset.h:
```diff
@@ -0,0 +1,198 @@
+// Get the next free (set to 0) index in the bitset.
+// It also sets the bit which it maybe should do... It really shouldn't do two things
+// But it's such a fast operation idk. // TODO: decide what you wanna do
+int bitset_get_free_idx(Bitset *bitset);
```
I think I'll choose to rename this for now.

MeirGavish commented on include/list.h:
```diff
@@ -1,27 +1,179 @@
+ *
+ * @return A new @ref ListItr
+ */
+ListItr list_itr_new(const List* list);
```
Maybe "init"?

ricfehr3 replied:
Idk, "init" doesn't have to do with creating the structure. What about `_create()` for everything, `bitset`,`list`,`list_itr`?

ricfehr3 commented on source/game.c:
```diff
@@ -220,17 +222,18 @@ static int hand_selections = 0;
 // Keeping track of what Jokers are scored at each step
-static int joker_scored_index = 0;
-static int joker_round_end_index = 0;
+static ListItr _joker_scored_itr;
```
That's fair. I moved that initialization to right after the _owned_joker_list is declared.

ricfehr3 commented:
> Looks good, just a few small changes left.
> 
> Since concerns over list traversal performance keep coming up I do want to note that if it does prove to be a problem in the future maybe it will be worth it to change some of the lists back to array lists, maybe just pre-allocate a large enough maximum size for them so there's no need to re-allocate, it might be good for small lists. I know we've been working a lot on this, but I just want to note the thoughts that come up over these concerns.

I have no problem switching back to the old method. I think there's value in this method (especially for insertions and `list_swap()` as you said). However, it may be worth it to add an array list (I'm going to call it a **vector** though, similar to c++). 

If the linked list methods work, I would say there's a benefit to only having one list type (less places for problems and less overhead), but I'm more than down to re-implement some of this as a vector too.

ricfehr3 commented:
> Oh, and I want to note one last thing, I think what we will need to add to this API for #158 is a `list_swap()` function that will allow efficiently swapping elements in the list, rather than a `list_insert()` function. This can be done in a separate issue or as part of #158.

I think we should do this in another PR, this one is just getting too big IMO.

MeirGavish reviewed (approved):
Looks great!

As for changing back to vector, right now it doesn't feel like we need to, it doesn't seem too bad for the small lists, but I did want to note it as an option.
I think this list should be used by the deck and hand too when we need to make them more dynamic. There are many benefits for this list for them. For example we see that drawing cards is a critical operation and we want to avoid reallocations there. Also the deck acts pretty much as a stack. These are bigger lists but maybe we can optimize critical parts to avoid traversing them too much.

I will probably merge this soon. Also thinking of #218 but I think this will go first.

MeirGavish commented on source/game.c:
```diff
@@ -386,10 +422,10 @@ void add_joker(JokerObject *joker_object)
 {
     // TODO: Extract to on_joker_removed() callback
-    JokerObject* joker_object = list_get(jokers, joker_idx);
+    //JokerObject* joker_object = list_get_at_idx(&_owned_jokers_list, joker_id);
```
There are a few other places, namely in the row selection, but maybe that can be optimized.
